Trimmed upstream GPC patch for Helium
---
From 9f6110d690bed64b39bb85adf99cd5343924cdff Mon Sep 17 00:00:00 2001
From: Maciej Czarnecki <mcczarny@gmail.com>
Date: Thu, 23 Jul 2026 11:48:32 -0700
Subject: [PATCH] Implement support for Global Privacy Control signal

Adding Sec-GPC header to requests when corresponding setting is enabled
and provide navigator.globalPrivacyControl property via js. Enabled with
GlobalPrivacyControl feature

Co-authored-by: Maxim Nesterov <m-nesterov@yandex-team.ru>
Feature: https://chromestatus.com/feature/5137324344213504
Spec: https://privacycg.github.io/gpc-spec/
Bug: 40745270
Change-Id: I03819d38688e357b119349e812a2a2f73505488e
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7319383
Commit-Queue: Ari Chivukula <arichiv@chromium.org>
Reviewed-by: Jonathan Lee <jonathanjlee@google.com>
Reviewed-by: Jeremy Roman <jbroman@chromium.org>
Reviewed-by: Nidhi Jaju <nidhijaju@chromium.org>
Reviewed-by: Ari Chivukula <arichiv@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1667282}
---

--- a/content/browser/loader/browser_initiated_resource_request.cc
+++ b/content/browser/loader/browser_initiated_resource_request.cc
@@ -9,6 +9,7 @@
 #include "content/public/browser/content_browser_client.h"
 #include "content/public/common/content_client.h"
 #include "net/http/http_request_headers.h"
+#include "third_party/blink/public/common/global_privacy_control/global_privacy_control_util.h"
 #include "third_party/blink/public/common/loader/loader_constants.h"
 #include "third_party/blink/public/common/renderer_preferences/renderer_preferences.h"
 
@@ -31,6 +32,18 @@ void UpdateAdditionalHeadersForBrowserIn
     headers->SetHeaderIfMissing(blink::kDoNotTrackHeader, "1");
   }
 
+  // Set the GlobalPrivacyControl header if appropriate.
+  // https://w3c.github.io/gpc/#expression-format
+  if (blink::IsGlobalPrivacyControlEnabled()) {
+    if (should_update_existing_headers) {
+      headers->RemoveHeader(blink::kGlobalPrivacyControlHeader);
+    }
+    headers->SetHeaderIfMissing(blink::kGlobalPrivacyControlHeader, "1");
+    blink::MaybeRecordGlobalPrivacyControlSourceMetric(
+        is_for_worker_script ? blink::GPCSignalSourceType::kWorkerNavigation
+                             : blink::GPCSignalSourceType::kFrameNavigation);
+  }
+
   // TODO(crbug.com/40833603): WARNING: This bypasses the permissions policy.
   // Unfortunately, workers lack a permissions policy and to derive proper hints
   // https://github.com/w3c/webappsec-permissions-policy/issues/207.
--- a/content/browser/renderer_host/ipc_utils.cc
+++ b/content/browser/renderer_host/ipc_utils.cc
@@ -480,7 +480,8 @@ bool VerifyNavigationHeaders(RenderProce
         header.name() != net::HttpRequestHeaders::kContentType &&
         header.name() != net::HttpRequestHeaders::kUserAgent &&
         header.name() != net::HttpRequestHeaders::kSecPurpose &&
-        header.name() != net::HttpRequestHeaders::kDNT) {
+        header.name() != net::HttpRequestHeaders::kDNT &&
+        header.name() != net::HttpRequestHeaders::kSecGPC) {
       if (base::FeatureList::IsEnabled(
               features::kKillOnInvalidNavigationHeaders)) {
         SCOPED_CRASH_KEY_STRING64("Bug487795397", "invalid_header",
--- a/content/renderer/render_frame_impl.cc
+++ b/content/renderer/render_frame_impl.cc
@@ -155,6 +155,7 @@
 #include "third_party/blink/public/common/context_menu_data/context_menu_data.h"
 #include "third_party/blink/public/common/context_menu_data/untrustworthy_context_menu_params.h"
 #include "third_party/blink/public/common/features.h"
+#include "third_party/blink/public/common/global_privacy_control/global_privacy_control_util.h"
 #include "third_party/blink/public/common/input/web_keyboard_event.h"
 #include "third_party/blink/public/common/interest_group/ad_auction_constants.h"
 #include "third_party/blink/public/common/loader/loader_constants.h"
@@ -4596,7 +4597,12 @@ void RenderFrameImpl::FinalizeRequestInt
     request.SetHttpHeaderField(
         blink::WebString::FromUtf8(blink::kDoNotTrackHeader), "1");
   }
-
+  if (blink::IsGlobalPrivacyControlEnabled()) {
+    request.SetHttpHeaderField(
+        blink::WebString::FromUtf8(blink::kGlobalPrivacyControlHeader), "1");
+    blink::MaybeRecordGlobalPrivacyControlSourceMetric(
+        blink::GPCSignalSourceType::kSubresourceFetch);
+  }
   // The request's extra data may indicate that we should set a custom user
   // agent. This needs to be done here, after WebKit is through with setting the
   // user agent on its own.
--- a/net/http/http_request_headers.h
+++ b/net/http/http_request_headers.h
@@ -103,6 +103,7 @@ class NET_EXPORT HttpRequestHeaders {
   static constexpr char kProxyConnection[] = "Proxy-Connection";
   static constexpr char kRange[] = "Range";
   static constexpr char kReferer[] = "Referer";
+  static constexpr char kSecGPC[] = "Sec-GPC";
   static constexpr char kSecPurpose[] = "Sec-Purpose";
   static constexpr char kTransferEncoding[] = "Transfer-Encoding";
   static constexpr char kUpgradeInsecureRequests[] =
--- a/third_party/blink/common/BUILD.gn
+++ b/third_party/blink/common/BUILD.gn
@@ -159,6 +159,7 @@ source_set("common") {
     "frame/from_ad_state.cc",
     "frame/user_activation_state.cc",
     "frame/view_transition_state_mojom_traits.cc",
+    "global_privacy_control/global_privacy_control_util.cc",
     "indexeddb/indexed_db_default_mojom_traits.cc",
     "indexeddb/indexeddb_key.cc",
     "indexeddb/indexeddb_key_path.cc",
--- a/third_party/blink/common/features.cc
+++ b/third_party/blink/common/features.cc
@@ -1045,6 +1045,9 @@ BASE_FEATURE(kFrequencyCappingForLargeSt
 BASE_FEATURE(kFrequencyCappingForOverlayPopupDetection,
              base::FEATURE_ENABLED_BY_DEFAULT);
 
+BASE_FEATURE(kGlobalPrivacyControlAlwaysSample,
+             base::FEATURE_DISABLED_BY_DEFAULT);
+
 BASE_FEATURE(kGMSCoreEmoji, base::FEATURE_ENABLED_BY_DEFAULT);
 
 BASE_FEATURE_PARAM(std::string,
--- /dev/null
+++ b/third_party/blink/common/global_privacy_control/global_privacy_control_util.cc
@@ -0,0 +1,39 @@
+// Copyright 2026 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "third_party/blink/public/common/global_privacy_control/global_privacy_control_util.h"
+
+#include "base/feature_list.h"
+#include "base/metrics/histogram_macros.h"
+#include "base/rand_util.h"
+#include "third_party/blink/public/common/features.h"
+
+namespace blink {
+namespace {
+// UMA key for GPC source samples.
+constexpr char kGlobalPrivacyControlSourceHistogram[] =
+    "Network.GlobalPrivacyControlSource.Subsampled";
+
+// Subsampling probability for the GPC source histogram.
+inline constexpr double kGlobalPrivacyControlSourceHistogramSampleProbability =
+    0.001;
+}  // namespace
+
+bool IsGlobalPrivacyControlEnabled() {
+  return base::FeatureList::IsEnabled(
+      blink::features::kGlobalPrivacyControlForce);
+}
+
+void MaybeRecordGlobalPrivacyControlSourceMetric(
+    GPCSignalSourceType source_type) {
+  if (base::FeatureList::IsEnabled(
+          blink::features::kGlobalPrivacyControlAlwaysSample) ||
+      base::ShouldRecordSubsampledMetric(
+          kGlobalPrivacyControlSourceHistogramSampleProbability)) {
+    UMA_HISTOGRAM_ENUMERATION(kGlobalPrivacyControlSourceHistogram,
+                              source_type);
+  }
+}
+
+}  // namespace blink
--- a/third_party/blink/common/loader/loader_constants.cc
+++ b/third_party/blink/common/loader/loader_constants.cc
@@ -7,5 +7,6 @@
 namespace blink {
 
 const char kDoNotTrackHeader[] = "DNT";
+const char kGlobalPrivacyControlHeader[] = "Sec-GPC";
 
 }  // namespace blink
--- a/third_party/blink/public/common/BUILD.gn
+++ b/third_party/blink/public/common/BUILD.gn
@@ -162,6 +162,7 @@ source_set("headers") {
     "frame/user_activation_state.h",
     "frame/user_activation_update_source.h",
     "frame/view_transition_state.h",
+    "global_privacy_control/global_privacy_control_util.h",
     "history/session_history_constants.h",
     "indexeddb/indexed_db_default_mojom_traits.h",
     "indexeddb/indexeddb_key.h",
--- a/third_party/blink/public/common/features.h
+++ b/third_party/blink/public/common/features.h
@@ -740,6 +740,10 @@ BLINK_COMMON_EXPORT BASE_DECLARE_FEATURE
 BLINK_COMMON_EXPORT BASE_DECLARE_FEATURE(
     kFrequencyCappingForOverlayPopupDetection);
 
+// If enabled, disables subsampling of GlobalPrivacyControl histogram entries.
+// Used in tests that verifies sampling in Renderer thread.
+BLINK_COMMON_EXPORT BASE_DECLARE_FEATURE(kGlobalPrivacyControlAlwaysSample);
+
 BLINK_COMMON_EXPORT BASE_DECLARE_FEATURE(kGMSCoreEmoji);
 
 BLINK_COMMON_EXPORT BASE_DECLARE_FEATURE_PARAM(
--- /dev/null
+++ b/third_party/blink/public/common/global_privacy_control/global_privacy_control_util.h
@@ -0,0 +1,37 @@
+// Copyright 2026 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef THIRD_PARTY_BLINK_PUBLIC_COMMON_GLOBAL_PRIVACY_CONTROL_GLOBAL_PRIVACY_CONTROL_UTIL_H_
+#define THIRD_PARTY_BLINK_PUBLIC_COMMON_GLOBAL_PRIVACY_CONTROL_GLOBAL_PRIVACY_CONTROL_UTIL_H_
+
+#include <string_view>
+
+#include "third_party/blink/public/common/common_export.h"
+
+namespace blink {
+
+// These values are persisted to logs. Entries should not be renumbered and
+// numeric values should never be reused.
+//
+// LINT.IfChange(GPCSignalSourceType)
+enum class GPCSignalSourceType {
+  kWorkerNavigation = 0,
+  kFrameNavigation = 1,
+  kSubresourceFetch = 2,
+  kWorkerSubresourceFetch = 3,
+  kMaxValue = kWorkerSubresourceFetch,
+};
+// LINT.ThenChange(//tools/metrics/histograms/metadata/network/enums.xml:GPCSignalSourceType)
+
+// Returns true if the Global Privacy Control is enabled.
+BLINK_COMMON_EXPORT bool IsGlobalPrivacyControlEnabled();
+
+// Records the source of the GPC signal in a subsampled histogram.
+// The histogram is subsampled one out of one thousand times.
+BLINK_COMMON_EXPORT void MaybeRecordGlobalPrivacyControlSourceMetric(
+    GPCSignalSourceType source_type);
+
+}  // namespace blink
+
+#endif  // THIRD_PARTY_BLINK_PUBLIC_COMMON_GLOBAL_PRIVACY_CONTROL_GLOBAL_PRIVACY_CONTROL_UTIL_H_
--- a/third_party/blink/public/common/loader/loader_constants.h
+++ b/third_party/blink/public/common/loader/loader_constants.h
@@ -12,6 +12,10 @@ namespace blink {
 // HTTP header set in requests to indicate they should be marked DoNotTrack.
 BLINK_COMMON_EXPORT extern const char kDoNotTrackHeader[];
 
+// HTTP request header `Sec-GPC`, set to "1" when Global Privacy Control is
+// enabled for the request.
+BLINK_COMMON_EXPORT extern const char kGlobalPrivacyControlHeader[];
+
 // These values indicate the load progress constants shared between both
 // //content and //blink.
 //
--- a/third_party/blink/public/mojom/use_counter/metrics/web_feature.mojom
+++ b/third_party/blink/public/mojom/use_counter/metrics/web_feature.mojom
@@ -5258,6 +5258,8 @@ enum WebFeature {
   kV8RTC_StartDiagnosticLogging_Method = 5941,
   kDeclarativePerformanceObserver = 5966,
   kPrefetchAndPrerenderActivationBeacon = 5967,
+  kV8WorkerNavigator_GlobalPrivacyControl_AttributeGetter = 5971,
+  kV8Navigator_GlobalPrivacyControl_AttributeGetter = 5972,
 
   // Add new features immediately above this line. Don't change the existing
   // numbers of any item, and don't reuse removed slots. Also don't add extra
--- a/third_party/blink/renderer/bindings/idl_in_modules.gni
+++ b/third_party/blink/renderer/bindings/idl_in_modules.gni
@@ -344,6 +344,7 @@ static_idl_files_in_modules = [
   "//third_party/blink/renderer/modules/gamepad/gamepad_touch.idl",
   "//third_party/blink/renderer/modules/gamepad/navigator_gamepad.idl",
   "//third_party/blink/renderer/modules/gamepad/window_gamepad.idl",
+  "//third_party/blink/renderer/modules/global_privacy_control/navigator_global_privacy_control.idl",
   "//third_party/blink/renderer/modules/handwriting/handwriting_drawing.idl",
   "//third_party/blink/renderer/modules/handwriting/handwriting_drawing_segment.idl",
   "//third_party/blink/renderer/modules/handwriting/handwriting_hints.idl",
--- a/third_party/blink/renderer/modules/BUILD.gn
+++ b/third_party/blink/renderer/modules/BUILD.gn
@@ -93,6 +93,7 @@ component("modules") {
     "//third_party/blink/renderer/modules/document_metadata",
     "//third_party/blink/renderer/modules/document_picture_in_picture",
     "//third_party/blink/renderer/modules/donottrack",
+    "//third_party/blink/renderer/modules/global_privacy_control",
     "//third_party/blink/renderer/modules/encoding",
     "//third_party/blink/renderer/modules/encryptedmedia",
     "//third_party/blink/renderer/modules/eventsource",
--- /dev/null
+++ b/third_party/blink/renderer/modules/global_privacy_control/BUILD.gn
@@ -0,0 +1,12 @@
+# Copyright 2026 The Chromium Authors
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import("//third_party/blink/renderer/modules/modules.gni")
+
+blink_modules_sources("global_privacy_control") {
+  sources = [
+    "navigator_global_privacy_control.cc",
+    "navigator_global_privacy_control.h",
+  ]
+}
--- /dev/null
+++ b/third_party/blink/renderer/modules/global_privacy_control/DEPS
@@ -0,0 +1,4 @@
+include_rules = [
+  "-third_party/blink/renderer/modules",
+  "+third_party/blink/renderer/modules/global_privacy_control",
+]
--- /dev/null
+++ b/third_party/blink/renderer/modules/global_privacy_control/navigator_global_privacy_control.cc
@@ -0,0 +1,24 @@
+// Copyright 2026 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "third_party/blink/renderer/modules/global_privacy_control/navigator_global_privacy_control.h"
+
+#include "third_party/blink/public/common/global_privacy_control/global_privacy_control_util.h"
+#include "third_party/blink/renderer/core/execution_context/navigator_base.h"
+#include "third_party/blink/renderer/core/frame/local_dom_window.h"
+#include "third_party/blink/renderer/core/frame/local_frame.h"
+#include "third_party/blink/renderer/core/frame/local_frame_client.h"
+
+namespace blink {
+namespace NavigatorGlobalPrivacyControl {
+
+bool globalPrivacyControl(NavigatorBase& navigator) {
+  // TODO(crbug.com/40745270): Currently, the GPC signal is controlled by a
+  // feature flag, when a user facing setting is added, this should be modified
+  // to use frame cached value.
+  return IsGlobalPrivacyControlEnabled();
+}
+
+}  // namespace NavigatorGlobalPrivacyControl
+}  // namespace blink
--- /dev/null
+++ b/third_party/blink/renderer/modules/global_privacy_control/navigator_global_privacy_control.h
@@ -0,0 +1,15 @@
+// Copyright 2026 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef THIRD_PARTY_BLINK_RENDERER_MODULES_GLOBAL_PRIVACY_CONTROL_NAVIGATOR_GLOBAL_PRIVACY_CONTROL_H_
+#define THIRD_PARTY_BLINK_RENDERER_MODULES_GLOBAL_PRIVACY_CONTROL_NAVIGATOR_GLOBAL_PRIVACY_CONTROL_H_
+
+namespace blink {
+class NavigatorBase;
+namespace NavigatorGlobalPrivacyControl {
+bool globalPrivacyControl(NavigatorBase&);
+}  // namespace NavigatorGlobalPrivacyControl
+}  // namespace blink
+
+#endif  // THIRD_PARTY_BLINK_RENDERER_MODULES_GLOBAL_PRIVACY_CONTROL_NAVIGATOR_GLOBAL_PRIVACY_CONTROL_H_
--- /dev/null
+++ b/third_party/blink/renderer/modules/global_privacy_control/navigator_global_privacy_control.idl
@@ -0,0 +1,13 @@
+// Copyright 2026 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+[
+    ImplementedAs=NavigatorGlobalPrivacyControl,
+    RuntimeEnabled=GlobalPrivacyControl
+] interface mixin NavigatorGlobalPrivacyControl {
+    [Measure] readonly attribute boolean globalPrivacyControl;
+};
+
+Navigator includes NavigatorGlobalPrivacyControl;
+WorkerNavigator includes NavigatorGlobalPrivacyControl;
--- a/third_party/blink/renderer/modules/service_worker/web_service_worker_fetch_context_impl.cc
+++ b/third_party/blink/renderer/modules/service_worker/web_service_worker_fetch_context_impl.cc
@@ -14,6 +14,7 @@
 #include "net/cookies/site_for_cookies.h"
 #include "services/network/public/cpp/wrapper_shared_url_loader_factory.h"
 #include "services/network/public/mojom/fetch_api.mojom-shared.h"
+#include "third_party/blink/public/common/global_privacy_control/global_privacy_control_util.h"
 #include "third_party/blink/public/common/loader/loader_constants.h"
 #include "third_party/blink/public/mojom/fetch/fetch_api_request.mojom-blink.h"
 #include "third_party/blink/public/platform/url_loader_throttle_provider.h"
@@ -150,6 +151,12 @@ void WebServiceWorkerFetchContextImpl::F
   if (renderer_preferences_.enable_do_not_track) {
     request.SetHttpHeaderField(WebString::FromUtf8(kDoNotTrackHeader), "1");
   }
+  if (IsGlobalPrivacyControlEnabled()) {
+    request.SetHttpHeaderField(WebString::FromUtf8(kGlobalPrivacyControlHeader),
+                               "1");
+    blink::MaybeRecordGlobalPrivacyControlSourceMetric(
+        blink::GPCSignalSourceType::kWorkerSubresourceFetch);
+  }
   auto url_request_extra_data = base::MakeRefCounted<WebURLRequestExtraData>();
   url_request_extra_data->set_originated_from_service_worker(true);
 
--- a/third_party/blink/renderer/platform/loader/fetch/url_loader/dedicated_or_shared_worker_global_scope_context_impl.cc
+++ b/third_party/blink/renderer/platform/loader/fetch/url_loader/dedicated_or_shared_worker_global_scope_context_impl.cc
@@ -15,6 +15,7 @@
 #include "base/task/thread_pool.h"
 #include "services/network/public/cpp/resource_request.h"
 #include "services/network/public/cpp/wrapper_shared_url_loader_factory.h"
+#include "third_party/blink/public/common/global_privacy_control/global_privacy_control_util.h"
 #include "third_party/blink/public/common/loader/loader_constants.h"
 #include "third_party/blink/public/mojom/service_worker/service_worker_fetch_handler_bypass_option.mojom-blink-forward.h"
 #include "third_party/blink/public/mojom/service_worker/service_worker_object.mojom.h"
@@ -377,6 +378,12 @@ void DedicatedOrSharedWorkerGlobalScopeC
   if (renderer_preferences_.enable_do_not_track) {
     request.SetHttpHeaderField(WebString::FromUtf8(kDoNotTrackHeader), "1");
   }
+  if (IsGlobalPrivacyControlEnabled()) {
+    request.SetHttpHeaderField(WebString::FromUtf8(kGlobalPrivacyControlHeader),
+                               "1");
+    blink::MaybeRecordGlobalPrivacyControlSourceMetric(
+        blink::GPCSignalSourceType::kWorkerSubresourceFetch);
+  }
 
   auto url_request_extra_data = base::MakeRefCounted<WebURLRequestExtraData>();
   request.SetURLRequestExtraData(std::move(url_request_extra_data));
--- a/third_party/blink/renderer/platform/runtime_enabled_features.json5
+++ b/third_party/blink/renderer/platform/runtime_enabled_features.json5
@@ -3361,6 +3361,14 @@
       status: "stable",
     },
     {
+      name: "GlobalPrivacyControl",
+      implied_by: ["GlobalPrivacyControlForce"],
+      base_feature: "none",
+    },
+    {
+      name: "GlobalPrivacyControlForce",
+    },
+    {
       name: "GraphemeClusterBoundsCheck",
       status: "stable",
     },
--- a/tools/metrics/histograms/enums.xml
+++ b/tools/metrics/histograms/enums.xml
@@ -17699,6 +17699,7 @@ from previous Chrome versions.
   <int value="1016132283" label="SeparateLocalAndAccountSearchEngines:enabled"/>
   <int value="1017364362" label="VrIconInDaydreamHome:enabled"/>
   <int value="1017523412" label="ExtensionsMenuAccessControl:disabled"/>
+  <int value="1017563501" label="GlobalPrivacyControlForce:enabled"/>
   <int value="1017617938"
       label="AutofillEnableBuyNowPayLaterForKlarna:disabled"/>
   <int value="1018101051" label="TabGroupShortcuts:enabled"/>
@@ -21127,6 +21128,7 @@ from previous Chrome versions.
   <int value="2101124624" label="WebUIDownloadShelf:disabled"/>
   <int value="2101151142" label="disable-direct-write"/>
   <int value="2101244145" label="NtpDoodleMurals:disabled"/>
+  <int value="2102392080" label="GlobalPrivacyControlForce:disabled"/>
   <int value="2102772909"
       label="EnableGenericOidcAuthProfileManagement:disabled"/>
   <int value="2103800063"
--- a/tools/metrics/histograms/metadata/blink/enums.xml
+++ b/tools/metrics/histograms/metadata/blink/enums.xml
@@ -6690,6 +6690,9 @@ Called by update_use_counter_feature_enu
   <int value="5941" label="V8RTC_StartDiagnosticLogging_Method"/>
   <int value="5966" label="DeclarativePerformanceObserver"/>
   <int value="5967" label="PrefetchAndPrerenderActivationBeacon"/>
+  <int value="5971"
+      label="V8WorkerNavigator_GlobalPrivacyControl_AttributeGetter"/>
+  <int value="5972" label="V8Navigator_GlobalPrivacyControl_AttributeGetter"/>
 </enum>
 
 <!-- LINT.ThenChange(//third_party/blink/public/mojom/use_counter/metrics/web_feature.mojom:WebFeature) -->
--- a/tools/metrics/histograms/metadata/network/enums.xml
+++ b/tools/metrics/histograms/metadata/network/enums.xml
@@ -432,6 +432,17 @@ chromium-metrics-reviews@google.com.
   <int value="1" label="Failure"/>
 </enum>
 
+<!-- LINT.IfChange(GPCSignalSourceType) -->
+
+<enum name="GPCSignalSourceType">
+  <int value="0" label="WorkerNavigation"/>
+  <int value="1" label="FrameNavigation"/>
+  <int value="2" label="SubresourceFetch"/>
+  <int value="3" label="WorkerSubresourceFetch"/>
+</enum>
+
+<!-- LINT.ThenChange(//third_party/blink/public/common/global_privacy_control/global_privacy_control_util.h:GPCSignalSourceType) -->
+
 <enum name="HermesResponseStatus">
   <summary>
     Statuses returned by Hermes, the Chrome OS daemon process which configures
--- a/tools/metrics/histograms/metadata/network/histograms.xml
+++ b/tools/metrics/histograms/metadata/network/histograms.xml
@@ -2152,6 +2152,20 @@ chromium-metrics-reviews@google.com.
   </token>
 </histogram>
 
+<histogram name="Network.GlobalPrivacyControlSource.Subsampled"
+    enum="GPCSignalSourceType" expires_after="2026-12-31">
+  <owner>arichiv@chromium.org</owner>
+  <owner>zalmashni@google.com</owner>
+  <summary>
+    During request initialization, this metric records the source of each
+    outgoing HTTP(S) request that had the Sec-GPC header set, categorized by the
+    type of the request.
+
+    Logged each time the Sec-GPC header is attached to an outgoing request,
+    during request finalization. Subsampled to one out of one thousand times.
+  </summary>
+</histogram>
+
 <histogram name="Network.Mdns.{ServiceType}" enum="MdnsQueryType"
     expires_after="2026-12-31">
   <owner>nmuggli@google.com</owner>
